Fix some SourceKit assertion hits

nathawes opened the pull request:
SourceKit's purely syntactic requests don't install a type-checker and so were hitting the below assertion in `ValueDecl::getInterfaceType()` when they inadvertently invoked some logic that required a type-checker:
```
assert(ctx.getLegacyGlobalTypeChecker()
         && "The type checker must be installed to make semantic queries!");
```

This was happening in calls to:
1) `FuncDecl::isStatic()`, which was calling `getInterfaceType()` unnecessarily in order to print it in a diagnostic message.
2) `ASTVerifier::verifyParsed()` for initializers and deinitializers which both call `isInvalid()` when the `init`/`deinit` isn't contained in a type/class decl. I just removed these checks since they're already being checked in `ASTVerifier::verifyChecked()` as well. 
3) `ValueDecl::getFormalAccess()` to provide the access level in the document structure response. ~~We could avoid this by approximating the access level syntactically, and actually used to do this before getFormalAccess was requestified, but the access level info doesn't seem like it's being used anywhere (at least not in Xcode or Swift Playgrounds) so I just removed it.~~ ~~This method was calling `isInvalid()` too (which I've replaced with what is hopefully an equivalent check), and also hit problems in inactive ifconfig clauses. It asserts that extensions are already bound and they're not in inactive ifconfig clauses (ASTScope doesn't support this at the moment).~~ This method calls `isInvalid()` too (though doesn't need to) but also triggers a different assertion if extensions aren't bound, and we don't want to do that here for performance reasons (these requests run on every keypress in the editor). I've resurrected the old `inferAccessLevel()` logic we used to use prior to `getFormalAccess()`.

Resolves rdar://problem/57202584

Commit:
[Sema] Avoid needing the typechecker for the IsStaticRequest request

It was being used purely to get the name of the type context for a diagnostic
message. SourceKit's syntactic-only requests were hitting an assertion when
this diagnostic was triggered because they don't set up a type checker.

Commit:
[ASTVerifier] Don't perform semantic requests in verifyParsed()

SourceKit doesn't install a typechecker for its syntactic requests so was
hitting assertions in these methods if any typechecking was triggered.

nathawes commented:
@swift-ci please test

johnfairh commented:
The `key.accessibility` has been present since the very early days and is relied on by widely-used open-source tools including at least SwiftLint and Jazzy -- it's a big problem to delete it.

Ideally keep the existing behaviour, but restoring the older syntactic algorithm would be OK --- though there was some fallout for us when this last changed.

@marcelofabri fyi

CodaFi commented on lib/AST/AccessRequests.cpp:
```diff
@@ -81,7 +81,7 @@ AccessLevelRequest::evaluate(Evaluator &evaluator, ValueDecl *D) const {
   // Special case for dtors and enum elements: inherit from container
   if (D->getKind() == DeclKind::Destructor ||
       D->getKind() == DeclKind::EnumElement) {
-    if (D->hasInterfaceType() && D->isInvalid()) {
```
I kinda wonder why this path has to exist at all.  I bet if you drop it you could preserve the “syntactic algorithm” that @johnfairh is asking for.

CodaFi replied:
I think what this means to ask is “is this enum element or destructor actually nested in a nominal type”.

CodaFi commented:
@johnfairh The problem is that the result of this key would not have been stable. There wasn’t a previous syntactic algorithm, there was only a quasi-semantic algorithm that used to get silently bounced out of the type checker with a null answer and whatever fallback that path gave you.

johnfairh commented:
Right; I think the 'previous syntactic' version that Nathan mentioned is the `inferAccessLevel()` stuff removed by c51f884879647bbee4a3de9c98503d9572f00caf, which was not always technically correct but close enough in practice.

marcelofabri commented:
[16 SwiftLint rules](https://github.com/realm/SwiftLint/search?p=1&q=accessibility&unscoped_q=accessibility) use this, as well as Jazzy.

Even though this key would not be stable, I don't remember ever seeing any false positives caused by this. cc @jpsim

nathawes commented:
Ah ok, thanks @johnfairh! I'll see if @CodaFi's suggestion works because that will probably have less change in behavior and go back to the `inferAccessLevel` logic if I hit problems.

CodaFi commented:
The type checker fixes in this round of changers LGTM

nathawes commented:
@swift-ci please test

CodaFi reviewed (approved):
Formalizing this: TypeChecker parts LGTM.  Can't speak for the syntax model.

Commit:
[IDE] Avoid name binding in sourcekitd's syntactic requests

It looks like we recently started binding extensions to their nominals in order
to continue to compute access levels via ValueDecl::getFormalAccess() after an
assertion was added to enforce that bindExtensions had been called before
anything tried to call ExtensionDecl::getBoundNominal() - which
getFormalAccess() depends on. Sourcekitd's syntactic requests are made on every
keypress in the editor though, so we shouldn't do any name binding (which may
require module loading) to keep them as fast as possible.

This patch restores the old inferAccessLevel() functions we used prior to the
switch to ValueDecl::getFormalAccess() (plus a few fixes) that does as much as
it can syntactically, without any name binding, and simply doesn't report the
access level in cases where it couldn't be computed without name-binding.

This also fixes an assertion hit we were getting trying to bind extensions in
inactive ifconfig clauses, which ASTScope doesn't support.

Resolves rdar://problem/57202584

nathawes commented:
Ok, so after thinking about this a bit more I don't think we can keep using `ValueDecl::getFormalAccess()` after all. To use it we need to (and recently started to) bind extensions to avoid hitting the reasonable assertion in `ExtensionDecl::getBoundNominal()` that extensions have actually been bound when `getFormalAccess()` calls it. The commit that added that assertion ([here](https://github.com/apple/swift/commit/4bc40303388691e33716eaf70d24ae568c9cf432#diff-07a8315a4aec7afc154b3b0dd52c5a43R797)) also modfied the Sourcekitd's syntactic requests to bind extensions to avoid tripping it. These requests are intended to be run on every keypress in the editor though, so we shouldn't be doing name lookup (which may require module loading) to keep them as fast as possible.

The newest commit stops binding extensions and restores the equivalent of the old `inferAccessLevel()` functions we used to have prior to the switch to `ValueDecl::getFormalAccess()` but with a few fixes to bring them back in sync with `getFormalAccess()`. It does as much as it can syntactically, and simply doesn't report the access level in cases where it couldn't be computed without name-binding.

@slavapestov are you ok with this? Or should I try to parameterize getFormalAccess() to have a 'syntactic-only' mode or something to share the code? (Just checking since you deduped this in the first place).

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

johnfairh commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1120,6 +1120,73 @@ static UIdent getAccessLevelUID(AccessLevel Access) {
+  if (D->getKind() == DeclKind::Destructor ||
+      D->getKind() == DeclKind::EnumElement) {
+    if (auto container = dyn_cast<NominalTypeDecl>(D->getDeclContext()))
+      return std::max(container->getFormalAccess(), AccessLevel::Internal);
```
Should this be calling `inferAccessSyntactically()` again instead of `getFormalAccess()`?

nathawes replied:
Yes! Thanks – I'll add a test case and fix.

Commit:
[IDE] Remove call to getFormalAccess() in inferAccessSyntactically()

It should have been recursing into inferAccessSyntactically() to avoid
any chance of triggering name lookup.

nathawes commented:
@swift-ci please test

nathawes commented:
Going to go ahead and merge this to let SourceKit stress tester CI job progress a little further, but if there are any further comments let me know and I'll address them in a follow-up PR.

slavapestov reviewed (commented):
I think this is fine in principle, but I'm wondering why do you need the access level at all for 'syntactic' requests?

nathawes commented:
Ideally we wouldn't. I'd like to remove it, and did in an earlier version of this PR but it's been in the SourceKit response since Swift was open sourced and from what others have said earlier in the comments several open source tools like SwiftLint and Jazzy currently rely on it. Longer term I think it'd be good if we could remove these syntactic requests completely in favor of using SwiftSyntax.

jpsim commented:
Those open source tools can certainly evolve and use preferred mechanisms to access that information. That migration path isn't completely clear to me, especially as SwiftSyntax grows its capabilities to match everything SourceKit's syntactic requests can provide.
